File: bindings/kotlin/src/Store.kt
// An in-memory conformant store, ported line-for-line from the Python
// binding's CURRENT store.py.
//
// Implements the store side of the abstract operation set (spec/store.md):
// immutable content objects with idempotent put; signed, add-only provenance
// records with a "quarantined" holding pen for unverifiable writes;
// materialized enrichment views with contributors (deduplicated by canonical
// entry); retraction handling in default views (history views carry a
// "retracted" flag); succession lineage; the resolve minimum
// (label-before-alias); the deterministic cycle-breaking view rule (exclude
// the cycle-completing record with the greatest (timestamp, id)); and the
// stigmergy gap read with its five gap kinds.
package org.causalontology

val CONTENT_KINDS = setOf("occurrent", "cro", "continuant", "realizable")
val RECORD_KINDS = setOf("assertion", "enrichment", "retraction", "succession")

// An enforcing store refused a write, with the reason as the message.
class RejectedWrite(message: String) : Exception(message)

class InMemoryStore(val enforcing: Boolean = true) {

    val objects = LinkedHashMap<String, JObj>()     // id -> content object
    val records = LinkedHashMap<String, JObj>()     // id -> provenance record
    val quarantine = LinkedHashMap<String, JObj>()  // id -> record (unsigned / unverifiable)

    // ------------------------------------------------------------------ put
    // Write a content object; idempotent; returns the identifier.
    fun put(objIn: JObj, kind: String? = null): String {
        val k = kind ?: Canonical.inferKind(objIn)
        if (k !in CONTENT_KINDS) {
            throw IllegalArgumentException("put() takes content objects; use putRecord()")
        }
        val obj = LinkedHashMap(objIn)
        if (!obj.containsKey("type")) obj["type"] = k
        if (!obj.containsKey("id")) obj["id"] = Canonical.identify(obj, k)
        val id = obj["id"] as String
        if (objects.containsKey(id)) {
            return id  // immutable: identical identity is a no-op
        }
        val (okSchema, whySchema) = Schema.validateSchema(obj, k)
        if (!okSchema) throw RejectedWrite(whySchema.joinToString("; "))
        val (okSem, whySem) = Semantics.validateSemantics(obj, k)
        if (!okSem) throw RejectedWrite(whySem.joinToString("; "))
        objects[id] = obj
        return id
    }

    // Write a signed provenance record; returns the identifier.
    fun putRecord(recordIn: JObj, kind: String? = null, force: Boolean = false): String {
        val k = kind ?: Canonical.inferKind(recordIn)
        if (k !in RECORD_KINDS) {
            throw IllegalArgumentException("putRecord() takes provenance records")
        }
        val record = LinkedHashMap(recordIn)
        if (!record.containsKey("type")) record["type"] = k
        val rid = (record["id"] as? String)?.takeIf { it.isNotEmpty() }
            ?: Canonical.identify(record, k)
        record["id"] = rid
        if (records.containsKey(rid)) {
            return rid  // add-only and idempotent
        }
        if (!Signing.verifyRecord(record, k)) {
            quarantine[rid] = record
            throw RejectedWrite("unsigned or unverifiable record: quarantined")
        }
        val (ok, why) = Semantics.validateSemantics(record, k)
        if (!ok) throw RejectedWrite(why.joinToString("; "))
        if (k == "retraction" && !retractionSourceOk(record)) {
            throw RejectedWrite(
                "a retraction is valid only from the retracted record's " +
                "source or its succession lineage")
        }
        if (k == "enrichment" && enforcing && !force) {
            val field = record["field"]
            if ((field == "subsumes" || field == "part_of") && wouldCycle(record)) {
                throw RejectedWrite(
                    "would create a cycle in the materialized $field graph")
            }
        }
        records[rid] = record
        return rid
    }

    // Simulate a decentralized replica merge (no enforcement gate).
    fun forceMergeRecord(record: JObj, kind: String? = null): String =
        putRecord(record, kind, force = true)

    // ------------------------------------------------------- record queries
    private fun recordsOf(kind: String): List<JObj> =
        records.values.filter { it["type"] == kind }

    private fun retractedIds(): Set<String> =
        recordsOf("retraction").map { it["retracts"] as String }.toSet()

    private fun retractionSourceOk(retraction: JObj): Boolean {
        val target = records[retraction["retracts"]]
            ?: return true  // open world: the target may arrive later
        return retraction["source"] in lineage(target["source"] as String)
    }

    // The succession chain closure containing key (includes key).
    fun lineage(key: String): Set<String> {
        val succ = HashMap<String, String>()
        val pred = HashMap<String, String>()
        for (s in recordsOf("succession")) {
            succ[s["predecessor"] as String] = s["successor"] as String
            pred[s["successor"] as String] = s["predecessor"] as String
        }
        val chain = mutableSetOf(key)
        var cursor = key
        while (pred.containsKey(cursor)) {
            cursor = pred[cursor]!!
            if (cursor in chain) break  // guard against pathological loops
            chain.add(cursor)
        }
        cursor = key
        while (succ.containsKey(cursor)) {
            cursor = succ[cursor]!!
            if (cursor in chain) break
            chain.add(cursor)
        }
        return chain
    }

    fun assertionsAbout(identifier: String, includeRetracted: Boolean = false): List<JObj> {
        val retracted = retractedIds()
        val out = mutableListOf<JObj>()
        for (r in recordsOf("assertion")) {
            if (r["about"] != identifier) continue
            if (r["id"] in retracted) {
                if (includeRetracted) {
                    val flagged = LinkedHashMap(r)
                    flagged["retracted"] = true
                    out.add(flagged)
                }
                continue
            }
            out.add(r)
        }
        return out
    }

    fun enrichmentsAbout(identifier: String, includeRetracted: Boolean = false): List<JObj> {
        val retracted = retractedIds()
        val out = mutableListOf<JObj>()
        for (r in recordsOf("enrichment")) {
            if (r["about"] != identifier) continue
            if (r["id"] in retracted && !includeRetracted) continue
            out.add(r)
        }
        return out
    }

    // ------------------------------------------------- materialized views
    // (active, excluded) for subsumes/part_of after rule 13 cycle-breaking.
    fun activeTaxonomyEdges(field: String): Pair<List<JObj>, List<JObj>> {
        val retracted = retractedIds()
        val recs = recordsOf("enrichment")
            .filter { it["field"] == field && it["id"] !in retracted }
        val active = recs.toMutableList()
        val excluded = mutableListOf<JObj>()
        while (true) {
            val cyc = findCycleRecords(active)
            if (cyc.isEmpty()) break
            // Exclude the cycle-completing record with the LATEST timestamp,
            // ties broken by lexicographic record identifier (deterministic).
            var loser = cyc[0]
            for (r in cyc) {
                val cmpTs = (r["timestamp"] as String).compareTo(loser["timestamp"] as String)
                val cmpId = (r["id"] as String).compareTo(loser["id"] as String)
                if (cmpTs > 0 || (cmpTs == 0 && cmpId > 0)) loser = r
            }
            active.remove(loser)
            excluded.add(loser)
        }
        return Pair(active, excluded)
    }

    // A depth-first search that returns the records along a path into a cycle
    // (mirrors store.py's _find_cycle_records, including iteration order).
    private fun findCycleRecords(recs: List<JObj>): List<JObj> {
        val edges = LinkedHashMap<String, MutableList<Pair<String, JObj>>>()
        for (r in recs) {
            edges.getOrPut(r["about"] as String) { mutableListOf() }
                .add(Pair(r["entry"] as String, r))
        }
        val state = HashMap<String, Int>()
        val cycle = mutableListOf<JObj>()

        fun dfs(node: String, pathRecords: List<JObj>): Boolean {
            state[node] = 1
            for ((nxt, rec) in edges[node] ?: emptyList()) {
                if (state.getOrElse(nxt) { 0 } == 1) {
                    cycle.addAll(pathRecords + rec)
                    return true
                }
                if (state.getOrElse(nxt) { 0 } == 0) {
                    if (dfs(nxt, pathRecords + rec)) return true
                }
            }
            state[node] = 2
            return false
        }

        for (start in edges.keys.toList()) {
            if (state.getOrElse(start) { 0 } == 0 && dfs(start, emptyList())) {
                return cycle
            }
        }
        return emptyList()
    }

    private fun wouldCycle(record: JObj): Boolean {
        val retracted = retractedIds()
        val recs = recordsOf("enrichment")
            .filter { it["field"] == record["field"] && it["id"] !in retracted }
        return findCycleRecords(recs + listOf(record)).isNotEmpty()
    }

    // The object with its materialized enrichment sets and contributors.
    fun get(identifier: String, view: String = "default"): JObj? {
        val obj = objects[identifier] ?: return null
        val includeRetracted = (view == "history")
        val excludedIds = mutableSetOf<String>()
        for (field in listOf("subsumes", "part_of")) {
            val (_, excluded) = activeTaxonomyEdges(field)
            excluded.forEach { excludedIds.add(it["id"] as String) }
        }
        // field -> canonical entry key -> {entry, contributors}
        val fields = LinkedHashMap<String, LinkedHashMap<String, MutableMap<String, Any?>>>()
        for (rec in enrichmentsAbout(identifier, includeRetracted)) {
            if (rec["id"] in excludedIds && view != "history") continue
            // Contributors deduplicate by the canonical (RFC 8785) form of the
            // entry - the equivalent of Python's sorted-items tuple key.
            val entryKey = Jcs.serialize(rec["entry"])
            val slot = fields.getOrPut(rec["field"] as String) { LinkedHashMap() }
            val bucket = slot.getOrPut(entryKey) {
                linkedMapOf("entry" to rec["entry"], "contributors" to mutableListOf<JObj>())
            }
            @Suppress("UNCHECKED_CAST")
            (bucket["contributors"] as MutableList<JObj>).add(
                linkedMapOf("source" to rec["source"], "timestamp" to rec["timestamp"]))
        }
        val enrichments = LinkedHashMap<String, Any?>()
        for ((f, slot) in fields) enrichments[f] = slot.values.toList()
        if (view == "raw") {
            return linkedMapOf("object" to obj)
        }
        return linkedMapOf("object" to obj, "enrichments" to enrichments)
    }

    // -------------------------------------------------------------- resolve
    private fun canonLabel(text: String): String =
        text.trim().lowercase().split(Regex("\\s+")).filter { it.isNotEmpty() }
            .joinToString("_")

    private fun normAlias(text: String): String =
        text.split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ").lowercase()

    // The conformance minimum: exact label, then alias, then nothing.
    fun resolve(text: String, lang: String? = null): List<String> {
        val labelHits = mutableListOf<String>()
        val aliasHits = mutableListOf<String>()
        val wantedLabel = canonLabel(text)
        val wantedAlias = normAlias(text)
        val retracted = retractedIds()
        for ((oid, obj) in objects) {
            if (obj["type"] !in listOf("occurrent", "continuant")) continue
            if (obj["label"] == wantedLabel) {
                labelHits.add(oid)
                continue
            }
            for (rec in recordsOf("enrichment")) {
                if (rec["about"] != oid || rec["field"] != "aliases") continue
                if (rec["id"] in retracted) continue
                val entry = asObj(rec["entry"])
                if (lang != null && entry["lang"] != lang) continue
                if (normAlias(entry["text"] as? String ?: "") == wantedAlias) {
                    aliasHits.add(oid)
                    break
                }
            }
        }
        return labelHits + aliasHits
    }

    // ---------------------------------------------------------------- gaps
    // The stigmergy read. Gap kinds per spec/store.md.
    fun gaps(kind: String? = null): List<JObj> {
        val out = mutableListOf<JObj>()
        val refined = mutableSetOf<String>()
        for (obj in objects.values) {
            val refines = obj["refines"] as? String
            if (obj["type"] == "cro" && !refines.isNullOrEmpty()) {
                val parent = objects[refines]
                if (parent != null) {
                    val (ok, _) = Semantics.refinementValid(obj, parent)
                    if (ok) refined.add(parent["id"] as String)
                }
            }
        }
        for ((oid, obj) in objects) {
            if (obj["type"] != "cro") continue
            // missing_field: lacking the temporal window or the modality -
            // mechanism and context may legitimately stay unspecified forever
            // (empty_mechanism is its own kind; absent context = context-free).
            if ((!obj.containsKey("temporal") || !obj.containsKey("modality")) &&
                oid !in refined) {
                out.add(linkedMapOf(
                    "id" to oid, "kind" to "missing_field",
                    "missing" to Semantics.isPartial(obj).second))
            }
            val mech = obj["mechanism"]
            if (!obj.containsKey("mechanism") || (mech is List<*> && mech.isEmpty())) {
                if (oid !in refined) {
                    out.add(linkedMapOf("id" to oid, "kind" to "empty_mechanism"))
                }
            }
        }
        for (field in listOf("subsumes", "part_of")) {
            val (_, excluded) = activeTaxonomyEdges(field)
            for (rec in excluded) {
                out.add(linkedMapOf(
                    "id" to rec["id"], "kind" to "inconsistent_hierarchy",
                    "note" to ("excluded by the deterministic " +
                               "cycle-breaking view rule")))
            }
        }
        // dangling_reference: a reference to an object absent from the store -
        // the red link that says "this page is wanted".
        for ((oid, obj) in objects) {
            val refs = mutableListOf<String?>()
            if (obj["type"] == "cro") {
                for (key in listOf("causes", "effects", "context", "mechanism")) {
                    val vals = obj[key] as? List<*> ?: emptyList<Any?>()
                    for (v in vals) refs.add(v as? String)
                }
                val refines = obj["refines"] as? String
                if (!refines.isNullOrEmpty()) refs.add(refines)
            } else if (obj["type"] == "realizable") {
                refs.add(obj["bearer"] as? String)
            }
            for (ref in refs) {
                if (!ref.isNullOrEmpty() && !objects.containsKey(ref)) {
                    out.add(linkedMapOf(
                        "id" to oid, "kind" to "dangling_reference", "ref" to ref))
                }
            }
        }
        // conflict: pairs of claims satisfying the formal test (rule 6).
        val cros = objects.values.filter { it["type"] == "cro" }
        for (i in cros.indices) {
            for (j in i + 1 until cros.size) {
                if (Semantics.conflicts(cros[i], cros[j])) {
                    out.add(linkedMapOf(
                        "kind" to "conflict",
                        "a" to cros[i]["id"], "b" to cros[j]["id"]))
                }
            }
        }
        return if (kind == null) out else out.filter { it["kind"] == kind }
    }
}

File: bindings/dart/lib/store.dart
/// An in-memory conformant store.
///
/// Implements the store side of the abstract operation set (spec/store.md):
/// immutable content objects with idempotent put; signed, add-only
/// provenance records; materialized enrichment views with contributors;
/// retraction handling in default views; succession lineage; the resolve
/// minimum; the deterministic cycle-breaking view rule; and the stigmergy
/// gap read.
///
/// Dart's default Map literal is a LinkedHashMap, so `objects` and
/// `records` iterate in insertion order - deliberately mirroring the
/// iteration order of the Python reference (dicts preserve insertion
/// order), which the resolve and view orderings depend on.
library;

import 'canonical.dart';
import 'jcs.dart';
import 'schema.dart';
import 'semantics.dart';
import 'signing.dart';

const Set<String> contentKinds = {'occurrent', 'causal_relation_object', 'continuant', 'realizable'};
const Set<String> recordKinds = {'assertion', 'enrichment', 'retraction', 'succession'};

/// An enforcing store refused a write, with the reason as [message].
class RejectedWrite implements Exception {
  final String message;
  RejectedWrite(this.message);
  @override
  String toString() => message;
}

class InMemoryStore {
  final bool enforcing;
  final Map<String, Map<String, dynamic>> objects = {}; // id -> content object
  final Map<String, Map<String, dynamic>> records = {}; // id -> provenance record
  final Map<String, Map<String, dynamic>> quarantine = {}; // unverifiable records

  InMemoryStore([this.enforcing = true]);

  // ---------------------------------------------------------------- put
  /// Write a content object; idempotent; returns the identifier.
  String put(Map<String, dynamic> obj, [String? kind]) {
    final k = kind ?? inferKind(obj);
    if (!contentKinds.contains(k)) {
      throw ArgumentError('put() takes content objects; use putRecord()');
    }
    final copy = Map<String, dynamic>.from(obj);
    copy.putIfAbsent('type', () => k);
    if (!copy.containsKey('id')) {
      copy['id'] = identify(copy, k);
    }
    final id = copy['id'] as String;
    if (objects.containsKey(id)) {
      return id; // immutable: identical identity is a no-op
    }
    final (schemaOk, schemaWhy) = validateSchema(copy, k);
    if (!schemaOk) {
      throw RejectedWrite(schemaWhy.join('; '));
    }
    final (semOk, semWhy) = validateSemantics(copy, k);
    if (!semOk) {
      throw RejectedWrite(semWhy.join('; '));
    }
    objects[id] = copy;
    return id;
  }

  /// Write a signed provenance record; returns the identifier.
  String putRecord(Map<String, dynamic> record,
      [String? kind, bool force = false]) {
    final k = kind ?? inferKind(record);
    if (!recordKinds.contains(k)) {
      throw ArgumentError('putRecord() takes provenance records');
    }
    final copy = Map<String, dynamic>.from(record);
    copy.putIfAbsent('type', () => k);
    final rid = (copy['id'] as String?) ?? identify(copy, k);
    copy['id'] = rid;
    if (records.containsKey(rid)) {
      return rid; // add-only and idempotent
    }
    if (!verifyRecord(copy, k)) {
      quarantine[rid] = copy;
      throw RejectedWrite('unsigned or unverifiable record: quarantined');
    }
    final (semOk, semWhy) = validateSemantics(copy, k);
    if (!semOk) {
      throw RejectedWrite(semWhy.join('; '));
    }
    if (k == 'retraction' && !_retractionSourceOk(copy)) {
      throw RejectedWrite(
          "a retraction is valid only from the retracted record's "
          'source or its succession lineage');
    }
    if (k == 'enrichment' && enforcing && !force) {
      final field = copy['field'] as String;
      if ((field == 'subsumes' || field == 'part_of') && _wouldCycle(copy)) {
        throw RejectedWrite(
            'would create a cycle in the materialized $field graph');
      }
    }
    records[rid] = copy;
    return rid;
  }

  /// Simulate a decentralized replica merge (no enforcement gate).
  String forceMergeRecord(Map<String, dynamic> record, [String? kind]) {
    return putRecord(record, kind, true);
  }

  // ----------------------------------------------------- record queries
  List<Map<String, dynamic>> _recordsOf(String kind) {
    return records.values.where((r) => r['type'] == kind).toList();
  }

  Set<String> _retractedIds() {
    final out = <String>{};
    for (final r in _recordsOf('retraction')) {
      out.add(r['retracts'] as String);
    }
    return out;
  }

  bool _retractionSourceOk(Map<String, dynamic> retraction) {
    final target = records[retraction['retracts'] as String];
    if (target == null) {
      return true; // open world: the target may arrive later
    }
    return lineage(target['source'] as String)
        .contains(retraction['source'] as String);
  }

  /// The succession chain closure containing key (includes key).
  Set<String> lineage(String key) {
    final succ = <String, String>{};
    final pred = <String, String>{};
    for (final s in _recordsOf('succession')) {
      succ[s['predecessor'] as String] = s['successor'] as String;
      pred[s['successor'] as String] = s['predecessor'] as String;
    }
    final chain = <String>{key};
    var cursor = key;
    while (pred.containsKey(cursor)) {
      cursor = pred[cursor]!;
      chain.add(cursor);
    }
    cursor = key;
    while (succ.containsKey(cursor)) {
      cursor = succ[cursor]!;
      chain.add(cursor);
    }
    return chain;
  }

  List<Map<String, dynamic>> assertionsAbout(String identifier,
      [bool includeRetracted = false]) {
    final retracted = _retractedIds();
    final out = <Map<String, dynamic>>[];
    for (final r in _recordsOf('assertion')) {
      if (r['about'] != identifier) continue;
      if (retracted.contains(r['id'])) {
        if (includeRetracted) {
          out.add(Map<String, dynamic>.from(r)..['retracted'] = true);
        }
        continue;
      }
      out.add(r);
    }
    return out;
  }

  List<Map<String, dynamic>> enrichmentsAbout(String identifier,
      [bool includeRetracted = false]) {
    final retracted = _retractedIds();
    final out = <Map<String, dynamic>>[];
    for (final r in _recordsOf('enrichment')) {
      if (r['about'] != identifier) continue;
      if (retracted.contains(r['id']) && !includeRetracted) continue;
      out.add(r);
    }
    return out;
  }

  // ------------------------------------------------- materialized views
  /// (active, excluded) for subsumes/part_of after rule 13 cycle-breaking.
  (List<Map<String, dynamic>>, List<Map<String, dynamic>>)
      activeTaxonomyEdges(String field) {
    final retracted = _retractedIds();
    final recs = _recordsOf('enrichment')
        .where((r) => r['field'] == field && !retracted.contains(r['id']))
        .toList();
    final active = List<Map<String, dynamic>>.from(recs);
    final excluded = <Map<String, dynamic>>[];
    while (true) {
      final cyc = _findCycleRecords(active);
      if (cyc.isEmpty) break;
      // Exclude the cycle-completing record with the LATEST timestamp,
      // ties broken by lexicographic record identifier (deterministic).
      var loser = cyc.first;
      for (final r in cyc.skip(1)) {
        final cmp = (r['timestamp'] as String)
            .compareTo(loser['timestamp'] as String);
        if (cmp > 0 ||
            (cmp == 0 &&
                (r['id'] as String).compareTo(loser['id'] as String) > 0)) {
          loser = r;
        }
      }
      active.remove(loser);
      excluded.add(loser);
    }
    return (active, excluded);
  }

  static List<Map<String, dynamic>> _findCycleRecords(
      List<Map<String, dynamic>> recs) {
    final edges = <String, List<(String, Map<String, dynamic>)>>{};
    for (final r in recs) {
      edges
          .putIfAbsent(r['about'] as String, () => [])
          .add((r['entry'] as String, r));
    }
    final state = <String, int>{};
    final cycle = <Map<String, dynamic>>[];

    bool dfs(String node, List<Map<String, dynamic>> pathRecords) {
      state[node] = 1;
      for (final (nxt, rec)
          in edges[node] ?? const <(String, Map<String, dynamic>)>[]) {
        if ((state[nxt] ?? 0) == 1) {
          cycle
            ..addAll(pathRecords)
            ..add(rec);
          return true;
        }
        if ((state[nxt] ?? 0) == 0) {
          if (dfs(nxt, [...pathRecords, rec])) return true;
        }
      }
      state[node] = 2;
      return false;
    }

    for (final start in edges.keys.toList()) {
      if ((state[start] ?? 0) == 0 && dfs(start, [])) {
        return cycle;
      }
    }
    return [];
  }

  bool _wouldCycle(Map<String, dynamic> record) {
    final retracted = _retractedIds();
    final recs = _recordsOf('enrichment')
        .where((r) =>
            r['field'] == record['field'] && !retracted.contains(r['id']))
        .toList();
    return _findCycleRecords([...recs, record]).isNotEmpty;
  }

  /// The object with its materialized enrichment sets and contributors.
  Map<String, dynamic>? get(String identifier, [String view = 'default']) {
    final obj = objects[identifier];
    if (obj == null) return null;
    final includeRetracted = view == 'history';
    final excludedIds = <String>{};
    for (final field in ['subsumes', 'part_of']) {
      final (_, excluded) = activeTaxonomyEdges(field);
      excludedIds.addAll(excluded.map((r) => r['id'] as String));
    }
    // field -> canonical entry key -> {"entry": ..., "contributors": [...]}
    final fields = <String, Map<String, Map<String, dynamic>>>{};
    for (final rec in enrichmentsAbout(identifier, includeRetracted)) {
      if (excludedIds.contains(rec['id']) && view != 'history') continue;
      // Canonical-entry dedup: the RFC 8785 form of the entry (sorted keys)
      // plays the role of Python's sorted-items tuple.
      final entryKey = jcs(rec['entry']);
      final slot = fields.putIfAbsent(rec['field'] as String, () => {});
      final bucket = slot.putIfAbsent(entryKey,
          () => {'entry': rec['entry'], 'contributors': <Map<String, dynamic>>[]});
      (bucket['contributors'] as List).add({
        'source': rec['source'],
        'timestamp': rec['timestamp'],
      });
    }
    final enrichments = <String, List<Map<String, dynamic>>>{
      for (final e in fields.entries) e.key: e.value.values.toList(),
    };
    if (view == 'raw') {
      return {'object': obj};
    }
    return {'object': obj, 'enrichments': enrichments};
  }

  // ------------------------------------------------------------ resolve
  static String _canonLabel(String text) {
    final words = text
        .trim()
        .toLowerCase()
        .split(RegExp(r'\s+'))
        .where((w) => w.isNotEmpty);
    return words.join('_');
  }

  static String _normAlias(String text) {
    final words =
        text.split(RegExp(r'\s+')).where((w) => w.isNotEmpty);
    return words.join(' ').toLowerCase();
  }

  /// The conformance minimum: exact label, then alias, then nothing.
  List<String> resolve(String text, [String? lang]) {
    final labelHits = <String>[];
    final aliasHits = <String>[];
    final wantedLabel = _canonLabel(text);
    final wantedAlias = _normAlias(text);
    final retracted = _retractedIds();
    for (final entry in objects.entries) {
      final oid = entry.key;
      final obj = entry.value;
      if (obj['type'] != 'occurrent' && obj['type'] != 'continuant') continue;
      if (obj['label'] == wantedLabel) {
        labelHits.add(oid);
        continue;
      }
      for (final rec in _recordsOf('enrichment')) {
        if (rec['about'] != oid || rec['field'] != 'aliases') continue;
        if (retracted.contains(rec['id'])) continue;
        final e = rec['entry'] as Map;
        if (lang != null && e['lang'] != lang) continue;
        if (_normAlias((e['text'] as String?) ?? '') == wantedAlias) {
          aliasHits.add(oid);
          break;
        }
      }
    }
    return [...labelHits, ...aliasHits];
  }

  // --------------------------------------------------------------- gaps
  /// The stigmergy read. Gap kinds per spec/store.md.
  List<Map<String, dynamic>> gaps([String? kind]) {
    var out = <Map<String, dynamic>>[];
    final refined = <String>{};
    for (final obj in objects.values) {
      final refines = obj['refines'];
      if (obj['type'] == 'causal_relation_object' && refines is String && refines.isNotEmpty) {
        final parent = objects[refines];
        if (parent != null) {
          final (ok, _) = refinementValid(obj, parent);
          if (ok) {
            refined.add(parent['id'] as String);
          }
        }
      }
    }
    for (final entry in objects.entries) {
      final oid = entry.key;
      final obj = entry.value;
      if (obj['type'] != 'causal_relation_object') continue;
      // missing_field: lacking the temporal window or the modality -
      // mechanism and context may legitimately stay unspecified forever
      // (empty_mechanism is its own kind; absent context = context-free).
      if ((!obj.containsKey('temporal') || !obj.containsKey('modality')) &&
          !refined.contains(oid)) {
        out.add({'id': oid, 'kind': 'missing_field', 'missing': isPartial(obj).$2});
      }
      final mech = obj['mechanism'];
      if (!obj.containsKey('mechanism') || (mech is List && mech.isEmpty)) {
        if (!refined.contains(oid)) {
          out.add({'id': oid, 'kind': 'empty_mechanism'});
        }
      }
    }
    for (final field in ['subsumes', 'part_of']) {
      final (_, excluded) = activeTaxonomyEdges(field);
      for (final rec in excluded) {
        out.add({
          'id': rec['id'],
          'kind': 'inconsistent_hierarchy',
          'note': 'excluded by the deterministic cycle-breaking view rule',
        });
      }
    }
    // dangling_reference: a reference to an object absent from the store -
    // the red link that says "this page is wanted".
    for (final entry in objects.entries) {
      final oid = entry.key;
      final obj = entry.value;
      var refs = <String?>[];
      if (obj['type'] == 'causal_relation_object') {
        refs = [
          ...((obj['causes'] as List?) ?? const []).cast<String>(),
          ...((obj['effects'] as List?) ?? const []).cast<String>(),
          ...((obj['context'] as List?) ?? const []).cast<String>(),
          ...((obj['mechanism'] as List?) ?? const []).cast<String>(),
        ];
        final refines = obj['refines'];
        if (refines is String && refines.isNotEmpty) {
          refs.add(refines);
        }
      } else if (obj['type'] == 'realizable') {
        refs = [obj['bearer'] as String?];
      }
      for (final ref in refs) {
        if (ref != null && ref.isNotEmpty && !objects.containsKey(ref)) {
          out.add({'id': oid, 'kind': 'dangling_reference', 'ref': ref});
        }
      }
    }
    // conflict: pairs of claims satisfying the formal test (rule 6).
    final cros = objects.values.where((o) => o['type'] == 'causal_relation_object').toList();
    for (var i = 0; i < cros.length; i++) {
      for (var j = i + 1; j < cros.length; j++) {
        if (conflicts(cros[i], cros[j])) {
          out.add({'kind': 'conflict', 'a': cros[i]['id'], 'b': cros[j]['id']});
        }
      }
    }
    if (kind != null) {
      out = out.where((g) => g['kind'] == kind).toList();
    }
    return out;
  }
}
